### DISCUSSION TITLE
How to decrypt data via node js after encrypted via $security.encrypt ?

### DISCUSSION THREAD

**Discussion by ghost** [pocketbase/pocketbase#4530] [Q&A] [answered]
How to decrypt data via node js after encrypted via $security.encrypt ?

I tried:

```javascript
export const decrypt = (encrypted: string) => {
	let decipher = crypto.createDecipheriv('aes-256-ecb', SECRET_ENCRYPT_KEY, null);
	let decrypted = decipher.update(encrypted, 'base64', 'utf8');
	return decrypted + decipher.final('utf8');
};
```

but got:
`Provider routines::wrong final block length",`

**Comment by ganigeorgiev** [comment, marked as answer]
I usually avoid answering security related questions when the question lack details to understand the context, but in this case I'll make an exception since it is not documented how `$security.encrypt` works (_I'll update the method documentation with the next release_).

`$security.encrypt`/`$security.decrypt` methods are direct mapping to the their [`security.Encrypt`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.3/tools/security#Encrypt)/[`security.Decrypt`](https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.22.3/tools/security#Decrypt) Go equivalents. Under the hood they use a standard AES-256 with GCM cipher (with the default nonce and tag sizes provided by the Go `crypto/cipher` package).

I'm not that familiar with Node.js and its `crypto` package but one of the first search results pointed to https://gist.github.com/kendru/965f46013352ba0a560fb0fa6c19d7ed. Below is an extracted and minimized version of the decode/decrypt method shown in the linked gist:

```js
import crypto from "crypto"

function decrypt(input, key) {
    const ivLength    = 12; // default nonce size used in Go
    const tagLength   = 16; // default tag size used in Go
    const inputBuffer = Buffer.from(input, "base64");
    const iv          = Buffer.allocUnsafe(ivLength);
    const tag         = Buffer.allocUnsafe(tagLength);
    const data        = Buffer.alloc(inputBuffer.length - ivLength - tagLength, 0);

    inputBuffer.copy(iv, 0, 0, ivLength);
    inputBuffer.copy(tag, 0, inputBuffer.length - tagLength);
    inputBuffer.copy(data, 0, ivLength);

    const decipher = crypto.createDecipheriv("aes-256-gcm", key, iv)

    decipher.setAuthTag(tag);

    return decipher.update(data, null, "utf8") + decipher.final("utf8");
}

// usage
try {
    const result = decrypt("YOUR_ENCRYPTED_TEXT", "YOUR_SECRET")
    console.log(result)
} catch (err) {
    // failed to decrypt...
}
```

If you need further help, please don't just "prompt" and try to elaborate on your use case to understand the broader picture, why this is needed in the first place and to avoid the needless back-and-forth.

**Reply by ghost** [reply]
Thank you. it works perfectly.
